### PR TITLE
[ASP.NET Core] Add `error.type` attribute for tracing and metrics

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #
Design discussion issue #

## Changes
Adds `error.type` attribute to trace and metrics.

https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md

`.NET8.0` implementation of metrics ignores setting `error.type` to status code value in case of [failure](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#:~:text=If%20response%20status%20code%20was%20sent%20or%20received%20and%20status%20indicates%20an%20error%20according%20to%20HTTP%20span%20status%20definition%2C%20error.type%20SHOULD%20be%20set%20to%20the%20status%20code%20number%20(represented%20as%20a%20string)) as per the spec. I have opened an issue to get this fixed https://github.com/dotnet/aspnetcore/issues/51620. For now, the issue is targeted to be fixed in `.NET9.0`.

The change in this PR ensures that the value is consistent with how `.NET8.0` is setting this attribute in metrics in order to avoid breaking change for users migrating from lower versions to `.NET8.0`.

This change **does not** resolve
https://github.com/open-telemetry/opentelemetry-dotnet/issues/4865

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 1]
```diff
@@ -32,9 +33,14 @@ internal sealed class HttpInMetricsListener : ListenerHandler
     internal const string HttpServerDurationMetricName = "http.server.duration";
     internal const string HttpServerRequestDurationMetricName = "http.server.request.duration";
 
+    internal const string OnUnhandledHostingExceptionEvent = "Microsoft.AspNetCore.Hosting.UnhandledException";
+    internal const string OnUnhandledDiagnosticsExceptionEvent = "Microsoft.AspNetCore.Diagnostics.UnhandledException";
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:37]: nit: these consts are repeated here and in the AspNetCoreMetrics class. Can these be consolidated into a single place?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 2]
```diff
@@ -66,23 +72,67 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
...
+            return;
         }
+
+        ctx.Items.Add(ErrorTypekey, exc.GetType().FullName);
+
+        // See https://github.com/dotnet/aspnetcore/blob/690d78279e940d267669f825aa6627b0d731f64c/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L252
+        // and https://github.com/dotnet/aspnetcore/blob/690d78279e940d267669f825aa6627b0d731f64c/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs#L174
+        // this makes sure that top-level properties on the payload object are always preserved.
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "The event source guarantees that top level properties are preserved")]
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:119]: This justification message is wrong. looks like it was copy/pasted from the EventSource class.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 3]
```diff
@@ -66,23 +72,67 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
...
+#endif
+        static bool TryFetchException(object payload, out Exception exc)
+            => ExceptionPropertyFetcher.TryFetch(payload, out exc) && exc != null;
+
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "The event source guarantees that top level properties are preserved")]
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:125]: Same

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 4]
```diff
@@ -66,23 +72,67 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
...
+#endif
+        static bool TryFetchHttpContext(object payload, out HttpContext ctx)
+            => HttpContextPropertyFetcher.TryFetch(payload, out ctx) && ctx != null;
     }
 
     public void OnEventWritten_Old(string name, object payload)
     {
         var context = payload as HttpContext;
+        context.Items.TryGetValue(name, out var item);
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:134]: what is this? i don't see `item` being used

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: good catch. this was unintended.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 5]
```diff
@@ -32,9 +33,14 @@ internal sealed class HttpInMetricsListener : ListenerHandler
...
     internal const string HttpServerRequestDurationMetricName = "http.server.request.duration";
 
+    internal const string OnUnhandledHostingExceptionEvent = "Microsoft.AspNetCore.Hosting.UnhandledException";
+    internal const string OnUnhandledDiagnosticsExceptionEvent = "Microsoft.AspNetCore.Diagnostics.UnhandledException";
     private const string OnStopEvent = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop";
     private const string EventName = "OnStopActivity";
     private const string NetworkProtocolName = "http";
+    private static readonly PropertyFetcher<Exception> ExceptionPropertyFetcher = new("Exception");
+    private static readonly PropertyFetcher<HttpContext> HttpContextPropertyFetcher = new("httpContext");
+    private static readonly object ErrorTypekey = "error.type";
```
Comment (TimothyMothra) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:43]: nit: CamelCase

```suggestion
    private static readonly object ErrorTypeKey = "error.type";
```

**Review by TimothyMothra — APPROVED** [review]
On the whole looks good. Left some minor comments.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 1 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:37]: we could, that would require changes in other places. better to handle it separately.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 2 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: fixed

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 6]
```diff
@@ -32,9 +33,14 @@ internal sealed class HttpInMetricsListener : ListenerHandler
     internal const string HttpServerDurationMetricName = "http.server.duration";
     internal const string HttpServerRequestDurationMetricName = "http.server.request.duration";
 
+    internal const string OnUnhandledHostingExceptionEvent = "Microsoft.AspNetCore.Hosting.UnhandledException";
+    internal const string OnUnhandledDiagnosticsExceptionEvent = "Microsoft.AspNetCore.Diagnostics.UnhandledException";
     private const string OnStopEvent = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop";
     private const string EventName = "OnStopActivity";
     private const string NetworkProtocolName = "http";
+    private static readonly PropertyFetcher<Exception> ExceptionPropertyFetcher = new("Exception");
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:41]: Should this be lower cased to `exception`? I'm assuming `PropertyFetcher` is case insensitive when retrieving the properties. https://github.com/dotnet/aspnetcore/blob/v7.0.13/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L299

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 7]
```diff
@@ -66,23 +72,66 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
...
-            if (this.emitNewAttributes)
-            {
-                this.OnEventWritten_New(name, payload);
-            }
+    public void OnExceptionEventWritten(string name, object payload)
+    {
+        // We need to use reflection here as the payload type is not a defined public type.
+        if (!TryFetchException(payload, out Exception exc) || !TryFetchHttpContext(payload, out HttpContext ctx))
+        {
+            AspNetCoreInstrumentationEventSource.Log.NullPayload(nameof(HttpInListener), nameof(this.OnExceptionEventWritten), HttpServerDurationMetricName);
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:109]:
```suggestion
            AspNetCoreInstrumentationEventSource.Log.NullPayload(nameof(HttpInMetricsListener), nameof(this.OnExceptionEventWritten), HttpServerDurationMetricName);
```

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 8]
```diff
@@ -66,23 +72,66 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
...
+            return;
         }
+
+        ctx.Items.Add(ErrorTypeKey, exc.GetType().FullName);
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:113]: Would the framework create an activity even if only add AspNetCore metrics instrumentation? If yes, then I think we could add this information to `Activity.Current`. Do you know of any obvious pros and cons to adding this information to `HttpContext` vs `Activity.Current`?

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 9]
```diff
@@ -520,6 +532,8 @@ private static List<MetricPoint> GetMetricPoints(Metric metric)
     private static KeyValuePair<string, object>[] AssertMetricPoint_New(
         MetricPoint metricPoint,
         string expectedRoute = "api/Values",
+        int expectedStatusCode = 200,
```
Comment (utpilla) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:535]: Could we get rid of the default values for the parameters?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 8 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: yes, the activity will always be there. Perf was the main reason to choose `HttpContext.Items`. However, we could use custom property as well on activity. Here are the results from benchmarking

Results when trace is also enabled and activity will have additional tags.

|                    Method |     Mean |    Error |   StdDev |   Median |   Gen0 | Allocated |
|-------------------------- |---------:|---------:|---------:|---------:|-------:|----------:|
|            UseActivityTag | 82.36 ns | 1.691 ns | 1.736 ns | 83.06 ns | 0.0085 |      40 B |
| UseActivityCustomProperty | 81.98 ns | 1.133 ns | 1.004 ns | 81.54 ns |      - |         - |
|           UseContextItems | 81.11 ns | 1.656 ns | 3.420 ns | 79.67 ns |      - |         - |


Results when only metrics will be enabled and activity will only have one tag

|                    Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------- |---------:|---------:|---------:|-------:|----------:|
|            UseActivityTag | 55.56 ns | 0.572 ns | 0.535 ns | 0.0085 |      40 B |
| UseActivityCustomProperty | 80.94 ns | 0.538 ns | 0.477 ns |      - |         - |
|           UseContextItems | 80.50 ns | 0.901 ns | 0.843 ns |      - |         - |

Based on above, I have updated the implementation to use custom property. We could follow the same thing on HttpClient side where we need to add `error.type`.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 6 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:41]: Yes, propertyFetcher is case insensitive.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: IIRC the first time you add a custom property to activity you pay for an allocation. Are we already adding one somewhere else? Seems a bit odd that the allocation isn't showing in the benchmark.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 10]
```diff
@@ -130,6 +171,7 @@ public void OnEventWritten_Old(string name, object payload)
 
     public void OnEventWritten_New(string name, object payload)
     {
+        var activity = Activity.Current;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:174]: Let's add a `Debug.Assert(activity != null)` here. I'm paranoid 🤣

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 10 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs]: we have a check [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/37481f1ee608a7781c2b4693b9bb5891bc767c11/src/Shared/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs#L46-L49)

**Review by utpilla — APPROVED** [review]

**Pull request merged**